src/transformers/modeling_utils.py

```diff
@@ -434,9 +434,13 @@ def load(module: nn.Module, prefix=""):
 
         for name, child in module._modules.items():
             if child is not None:
-                load(child, prefix + name + ".")
+                load(child, state_dict, prefix + name + ".")
 
-    load(model_to_load, prefix=start_prefix)
+    load(model_to_load, state_dict, prefix=start_prefix)
+    # Delete `state_dict` so it could be collected by GC earlier. Note that `state_dict` is a copy of the argument, so
+    # it's safe to delete it. We don't call `gc.collect` here, instead let GC make its own decision.
+    # See https://github.com/huggingface/transformers/issues/18782
```

Comment:
A bit big as a comment (we can't have three lines of comments for each new line of code ;-) )
```suggestion
    # Delete `state_dict` so it could be collected by GC earlier. Note that `state_dict` is a copy of the argument, so
    # it's safe to delete it.
```